radix-engine/src/kernel/call_frame.rs

```diff
@@ -511,12 +480,18 @@ impl<L: Clone> CallFrame<L> {
         Ok((lock_handle, substate_value.len(), store_access))
     }
 
-    pub fn close_substate<S: SubstateStore>(
+    pub fn close_substate<S: SubstateStore, F, M>(
         &mut self,
         heap: &mut Heap,
         store: &mut S,
+        callback: &mut M,
```

Comment:
Can we remove callback object from this layer? I think the kernel itself should be able to add callback object on it's own.

Reply:
I kind of need it inside the closure.

But we don't need `M` to be `KernelCallbackObject`. Maybe I just remove the type bound?

Another option is have trait object which encapsulates both the `callback` and the `filter`.

Reply:
Yes, exactly, the call_frame doesn't need to know that there is even such a thing as filter.

See a similar pattern here: https://github.com/radixdlt/radixdlt-scrypto/pull/1270/files#diff-d92ad4919aac15aa05fcea4887fe2e880ecd22d05a6b77dfd51479662372a17fL261

Reply:
right, that's what I meant